lib/ASTGen/Sources/ASTGen/ASTGen.swift

```diff
@@ -477,21 +510,155 @@ extension Optional where Wrapped: LazyCollectionProtocol {
 /// Generate AST nodes for all top-level entities in the given source file.
 @_cdecl("swift_ASTGen_buildTopLevelASTNodes")
 public func buildTopLevelASTNodes(
-  diagEnginePtr: UnsafeMutableRawPointer,
-  sourceFilePtr: UnsafePointer<UInt8>,
-  dc: UnsafeMutableRawPointer,
-  ctx: UnsafeMutableRawPointer,
+  diagEngine: BridgedDiagnosticEngine,
+  sourceFilePtr: UnsafeRawPointer,
+  dc: BridgedDeclContext,
+  ctx: BridgedASTContext,
+  legacyParser: BridgedLegacyParser,
   outputContext: UnsafeMutableRawPointer,
   callback: @convention(c) (UnsafeMutableRawPointer, UnsafeMutableRawPointer) -> Void
 ) {
-  sourceFilePtr.withMemoryRebound(to: ExportedSourceFile.self, capacity: 1) { sourceFile in
-    ASTGenVisitor(
-      diagnosticEngine: .init(raw: diagEnginePtr),
-      sourceBuffer: sourceFile.pointee.buffer,
-      declContext: BridgedDeclContext(raw: dc),
-      astContext: BridgedASTContext(raw: ctx)
+  let sourceFile = sourceFilePtr.assumingMemoryBound(to: ExportedSourceFile.self)
+  ASTGenVisitor(
+    diagnosticEngine: diagEngine,
+    sourceBuffer: sourceFile.pointee.buffer,
+    declContext: dc,
+    astContext: ctx,
+    legacyParser: legacyParser
+  )
+  .generate(sourceFile.pointee.syntax)
+  .forEach { callback($0, outputContext) }
+}
+
+/// Generate an AST node at the given source location. Returns the generated
+/// ASTNode and mutate the pointee of `endLocPtr` to the end of the node.
+private func _build<Node: SyntaxProtocol>(
+  kind: Node.Type,
+  diagEngine: BridgedDiagnosticEngine,
+  sourceFilePtr: UnsafeRawPointer,
+  sourceLoc: BridgedSourceLoc,
+  declContext: BridgedDeclContext,
+  astContext: BridgedASTContext,
+  legacyParser: BridgedLegacyParser,
+  endLocPtr: UnsafeMutablePointer<BridgedSourceLoc>
+) -> UnsafeMutableRawPointer? {
+  let sourceFile = sourceFilePtr.assumingMemoryBound(to: ExportedSourceFile.self)
+
+  // Find the type syntax node.
+  guard
+    let node = findSyntaxNodeInSourceFile(
+      sourceFilePtr: sourceFilePtr,
+      // FIXME: findSyntaxNodeInSourceFile should receive `BridgedSourceLoc`.
+      sourceLocationPtr: sourceLoc.getOpaquePointerValue()?.assumingMemoryBound(to: UInt8.self),
+      type: Node.self,
+      wantOutermost: true
     )
-    .generate(sourceFile.pointee.syntax)
-    .forEach { callback($0, outputContext) }
-  }
+  else {
+    // FIXME: Produce an error
+    return nil
```

Comment:
Currently `findSyntaxNodeInSourceFile()` prints an error message to stdout, which is not great, but at least it prevents "unwanted surprises". But yeah, I think we should add a diagnostic on the diagnostic engine.
But since this PR is getting bigger, and to unblock some follow-up patches I have, I'd like to do it in the followups.
Also, I'm thinking to change `findSyntaxNodeInSourceFile()` itself. https://github.com/apple/swift/pull/69724#discussion_r1387258659